### PR TITLE
Make client register with scheduler on creation

### LINKED ISSUE #340: Client should verify Scheduler's existence
On startup a Client should perform a handshake with the scheduler to ensure that it exists.  This would presumably be similar to the Worker-Scheduler relationship.

### PR DESCRIPTION
**Pull request by cowlicks**
closes #340

### CODE REVIEW

File: dask/distributed/client.py [hunk 1]
```diff
@@ -33,6 +33,7 @@ def __init__(self, scheduler, address=None):
         self.socket = context.socket(zmq.DEALER)
         self.socket.setsockopt(zmq.IDENTITY, self.address)
         self.socket.connect(self.address_to_scheduler)
+        self.send_to_scheduler({'function': 'register'}, {})
```
Comment (mrocklin) [inline dask/distributed/client.py]: Do we want to send any identifying state to the scheduler?  The second arugument here `{}` holds metadata about the client.

**Comment by mrocklin** [comment]
This could use a brief test

**Comment by mrocklin** [comment]
Looks good

**Comment by cowlicks** [comment]
Comments addressed, anything else?

File: dask/distributed/tests/test_client.py [hunk 2]
```diff
@@ -111,3 +113,9 @@ def test_register_collections():
         c.close()
         d.close()
 
+
+def test_register_with_scheduler():
+    with scheduler_and_workers() as (s, (a, b)):
+        c = Client(s.address_to_clients)
+        sleep(.2)
```
Comment (mrocklin) [inline dask/distributed/tests/test_client.py]:
```python
for i in range(100):
    if s.clients:
        break
    else:
        sleep(0.1)
```

A number of the intermittent errors are caused by `sleep`s.  I've been slowly removing them.

Alternatively the client could block until it receives an acknowledgment from the Scheduler.  This is probably cleaner.

Comment (cowlicks) [inline dask/distributed/tests/test_client.py]: Yeah I think the client blocking makes sense. I'll add that.

Comment (cowlicks) [inline dask/distributed/tests/test_client.py]: The client also gets a dict of the registered workers and clients when it registers now.

File: dask/distributed/scheduler.py [hunk 3]
```diff
@@ -209,6 +213,15 @@ def block(self):
         self._listen_to_workers_thread.join()
         self._listen_to_clients_thread.join()
 
+    def _client_registration(self, header, payload):
+        """ Client comes in, register it, send back info about the cluster"""
+        payload = pickle.loads(payload)
+        address = header['address']
+        self.clients[address] = payload
+        out_header = {}
+        out_payload = {'clients': self.clients, 'workers': self.workers}
```
Comment (mrocklin) [inline dask/distributed/scheduler.py]: would suggest removing `self.clients` from this.  Not sure we want to publicize clients to each other.

**Comment by mrocklin** [comment]
This seems good to me.  I think that at some point we'll want to separate the getting of worker information from client registration (one might want to get an updated view of the workers at some later time).  But for now I say we go with this.  +1 after tests pass.

**Pull request merged**